zio-http/src/main/scala/zio/http/service/ZKeyedPool.scala

```diff
@@ -0,0 +1,173 @@
+   * live before a pool whose excess items are not being used will be shrunk
+   * down to the minimum size. The pool is returned in a `Scope`, which governs
+   * the lifetime of the pool. When the pool is shutdown because the `Scope` is
+   * used, the individual items allocated by the pool will be released in some
+   * unspecified order.
+   */
+  def make[Key, Env: Tag, Err, Item](get: Key => ZIO[Env, Err, Item], range: => Range, timeToLive: => Duration)(implicit
+    trace: Trace,
+  ): ZIO[Env with Scope, Nothing, ZKeyedPool[Err, Key, Item]] =
+    make(get, _ => range, _ => timeToLive)
+
+  /**
+   * Makes a new pool with the specified minimum and maximum sizes and time to
+   * live before a pool whose excess items are not being used will be shrunk
+   * down to the minimum size. The pool is returned in a `Scope`, which governs
+   * the lifetime of the pool. When the pool is shutdown because the `Scope` is
+   * used, the individual items allocated by the pool will be released in some
+   * unspecified order.
+   *
+   * The size of the underlying pools can be configured per key.
+   */
+  def make[Key, Env: Tag, Err, Item](get: Key => ZIO[Env, Err, Item], range: Key => Range, timeToLive: Key => Duration)(
+    implicit trace: Trace,
+  ): ZIO[Env with Scope, Nothing, ZKeyedPool[Err, Key, Item]] =
+    makeWith(get, range)((key: Key) => Some(timeToLive(key)))
+
+  private def makeWith[Key, Env: Tag, Err, Item](get: Key => ZIO[Env, Err, Item], range: Key => Range)(
+    ttl: Key => Option[Duration],
+  )(implicit
+    trace: Trace,
+  ): ZIO[Env with Scope, Nothing, ZKeyedPool[Err, Key, Item]] =
+    for {
+      activePools   <- TMap.empty[Key, ZPool[Err, Item]].commit
+      createdPools  <- TSet.empty[Key].commit
+      acquiredItems <- TMap.empty[Item, Key].commit
+      env           <- ZIO.environment[Env]
+      createPool = (key: Key) =>
+        ttl(key) match {
+          case Some(timeToLive) =>
+            ZPool
+              .make(get(key), range(key), timeToLive)
+              .provideSomeEnvironment[Scope](_.union[Env](env))
+          case None             =>
+            ZPool
+              .make(get(key), range(key).end)
+              .provideSomeEnvironment[Scope](_.union[Env](env))
+        }
+    } yield DefaultKeyedPool[Err, Key, Item](
+      createPool,
+      activePools,
+      createdPools,
+      acquiredItems,
+    )
+
+  private final case class DefaultKeyedPool[Err, Key, Item](
+    createPool: Key => ZIO[Scope, Nothing, ZPool[Err, Item]],
+    activePools: TMap[Key, ZPool[Err, Item]],
+    createdPools: TSet[Key],
+    acquiredItems: TMap[Item, Key],
+  ) extends ZKeyedPool[Err, Key, Item] {
+
+    override def get(key: Key)(implicit trace: Trace): ZIO[Scope, Err, Item] =
+      createdPools
+        .contains(key)
+        .flatMap { alreadyCreated =>
+          if (alreadyCreated) {
+            activePools.get(key).flatMap {
+              case None       =>
+                ZSTM.retry
```

Comment:
What is this about?

Reply:
Creating a new pool is an effect so we can't do it as part of the transaction, so there is a race condition between checking if you need to create a new one, creating it and then storing it. 
So I chose to maintain a set beside the key->pool map because we can record the intention to create a new pool within the same transaction where we check if it already exists.

So either:
- the pool is already in the map and in the set, then we use it
- the key is in the set, but the pool is not in the map, so we retry because another fiber is already creating it
- not in the set and not in the map, then we put it in the set, commit, create the pool and put it in the map

I think this could be replaced by a key->promise[value] map instead, which may be simpler.